File: sample_apps/python/CsvIngestionExample.py
import csv
import time
from Constant import DATABASE_NAME, TABLE_NAME


class CsvIngestionExample:
    def __init__(self, client):
        self.client = client

    def bulk_write_records(self, filepath):
        with open(filepath, 'r') as csv_file:
            # creating a csv reader object
            csv_reader = csv.reader(csv_file)

            records = []
            current_time = self._current_milli_time()
            counter = 0

            # extracting each data row one by one
            for row in csv_reader:
                dimensions = [
                    {'Name': row[0], 'Value': row[1]},
                    {'Name': row[2], 'Value': row[3]},
                    {'Name': row[4], 'Value': row[5]}
                ]

                record_time = current_time - (counter * 50)

                record = {
                    'Dimensions': dimensions,
                    'MeasureName': row[6],
                    'MeasureValue': row[7],
                    'MeasureValueType': row[8],
                    'Time': str(record_time)
                }

                records.append(record)
                counter = counter + 1

                if len(records) == 100:
                    self._submit_batch(records, counter)
                    records = []

            if len(records) != 0:
                self._submit_batch(records, counter)

            print("Ingested %d records" % counter)

    def _submit_batch(self, records, counter):
        try:
            result = self.client.write_records(DatabaseName=DATABASE_NAME, TableName=TABLE_NAME,
                                               Records=records, CommonAttributes={})
            print("Processed [%d] records. WriteRecords Status: [%s]" % (counter,
                                                                         result['ResponseMetadata']['HTTPStatusCode']))
        except Exception as err:
            print("Error:", err)

    @staticmethod
    def _current_milli_time():
        return int(round(time.time() * 1000))

File: sample_apps/js/csv-ingestion-example.js

import fs from 'fs';
import readline from 'readline';
import { constants } from "./constants.js";
import { WriteRecordsCommand } from "@aws-sdk/client-timestream-write";

export async function processCSV(writeClient, filePath) {
    try {
        await ingestCsvRecords(writeClient, filePath);
    } catch (e) {
        console.log('e', e);
    }
}

async function ingestCsvRecords(writeClient, filePath) {
    const currentTime = Date.now().toString(); // Unix time in milliseconds

    var records = [];
    var counter = 0;

    const rl = readline.createInterface({
        input: fs.createReadStream(filePath),
        crlfDelay: Infinity
    });

    const promises = [];

    for await ( const dataRow of rl ) {
        var row = dataRow.toString().split(',');
        const dimensions = [
            {'Name': row[0].toString(), 'Value': row[1].toString()},
            {'Name': row[2].toString(), 'Value': row[3].toString()},
            {'Name': row[4].toString(), 'Value': row[5].toString()}
        ];
        const recordTime = currentTime - counter * 50;
        const record = {
            'Dimensions': dimensions,
            'MeasureName': row[6].toString(),
            'MeasureValue': row[7].toString(),
            'MeasureValueType': row[8].toString(),
            'Time': recordTime.toString()
        };

        records.push(record);
        counter++;

        if (records.length === 100) {
            promises.push(submitBatch(writeClient, records, counter));
            records = [];
        }
    }

    if (records.length !== 0) {
        promises.push(submitBatch(writeClient, records, counter));
    }

    await Promise.all(promises);

    console.log(`Ingested ${counter} records`);
}

function submitBatch(writeClient, records, counter) {
    const params = new WriteRecordsCommand({
        DatabaseName: constants.DATABASE_NAME,
        TableName: constants.TABLE_NAME,
        Records: records
    });

    return writeClient.send(params).then(
        (data) => {
            console.log(`Processed ${counter} records.`);
        },
        (err) => {
            console.log("Error writing records:", err);
        }
    );
}